### PR TITLE
Extract method `Select.escapeQuotes` to a separate public class Quotes

### PR DESCRIPTION
**Pull request by asolntsev**
... to enable reusing this code in other classes and frameworks.

Also improved code and tests.

### CODE REVIEW

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package org.openqa.selenium.support.ui;
+
+public class Quotes {
+
+  /**
+   * Convert strings with both quotes and ticks into: foo'"bar -> concat("foo'", '"', "bar")
```
Comment (jleyba) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: This sentence doesn't make sense. Converts strings into what?

**Comment by andreastt** [comment]
First of all, thanks for the patch!  What is the reason you want to extract this into its own class?

**Comment by asolntsev** [comment]
Actually it was a comment from the previous code, just moved to javadoc.

But I agree, it was incorrect. I just have fixed it.

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+   * Convert strings with both quotes and ticks into a valid xpath component "concat(...)"
+   * 
+   * For example, foo'"bar will be converted to concat("foo'", '"', "bar")
+   * 
+   * @param toEscape a text to escape quotes in, e.g. "f'oo"
+   * @return the same text with escaped quoted, e.g. "\"f'oo\""
+   */
+  public static String escape(String toEscape) {
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: I feel this should take the type of quote to replace as an argument.

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package org.openqa.selenium.support.ui;
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: Missing copyright boilerplate.

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+public class Quotes {
+
+  /**
+   * Convert strings with both quotes and ticks into a valid xpath component "concat(...)"
+   * 
+   * For example, foo'"bar will be converted to concat("foo'", '"', "bar")
+   * 
+   * @param toEscape a text to escape quotes in, e.g. "f'oo"
+   * @return the same text with escaped quoted, e.g. "\"f'oo\""
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: Wrap both of these in {@code …}

**Comment by asolntsev** [comment]
> What is the reason you want to extract this into its own class?
I want to use the same code in my tests. For example, for composing XPath that would search element by text. 

Actually it's also good for Selenium code: it decoupled the code and allowed to make smaller isolated tests for quotes escaping.

File: java/client/test/org/openqa/selenium/support/ui/QuotesTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2007-2009 Selenium committers
```
Comment (andreastt) [inline java/client/test/org/openqa/selenium/support/ui/QuotesTest.java]: Change this to:

```
Copyright 2015 Software Freedom Conservancy
Copyright 2007-2009 Selenium committers
```

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2007-2015 Selenium committers
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: Please add `Copyright 2015 Software Freedom Conservancy` to this as well.

**Comment by andreastt** [comment]
Generally I'm sceptical of too much abstraction but this seems to make the original string building code much easier to read.

File: java/client/src/org/openqa/selenium/support/ui/Quotes.java (hunk 1 above)
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/ui/Quotes.java]: This is still relevant, not sure why Github marked it as outdated.

**Comment by andreastt** [comment]
You also need to include the new files in the `build.desc` files in their respective directories, otherwise they won't be picked up by the build system.

**Comment by asolntsev** [comment]
Oh, is it really automated system? Build system in 21 century could automatically find and run all the tests...

Anyway, I have added QuotesTest to SmallTests suite. I hope it's enough.

**Comment by andreastt** [comment]
Your IDE can detect which tests are _runnable_, but not how to run them without you telling it to run it.  Because we need reproducible builds we also need to tell the build system where to find the source of what to compile so that it doesn't get slowed down by trying to compile everything in one big batch.

**Comment by andreastt** [comment]
```
Compiling: //java/client/src/org/openqa/selenium/support/ui:elements as build/java/client/src/org/openqa/selenium/support/ui/elements.jar
----------
1. ERROR in /Users/ato/Code/selenium/java/client/src/org/openqa/selenium/support/ui/Select.java (at line 115)
	element.findElements(By.xpath(".//option[normalize-space(.) = " + Quotes.escape(text) + "]"));
	                                                                  ^^^^^^
Quotes cannot be resolved
----------
2. ERROR in /Users/ato/Code/selenium/java/client/src/org/openqa/selenium/support/ui/Select.java (at line 136)
	Quotes.escape(subStringWithoutSpace) + ")]"));
	^^^^^^
Quotes cannot be resolved
----------
3. ERROR in /Users/ato/Code/selenium/java/client/src/org/openqa/selenium/support/ui/Select.java (at line 202)
	".//option[@value = " + Quotes.escape(value) + "]"));
	                        ^^^^^^
Quotes cannot be resolved
----------
4. ERROR in /Users/ato/Code/selenium/java/client/src/org/openqa/selenium/support/ui/Select.java (at line 247)
	".//option[@value = " + Quotes.escape(value) + "]"));
	                        ^^^^^^
Quotes cannot be resolved
----------
5. ERROR in /Users/ato/Code/selenium/java/client/src/org/openqa/selenium/support/ui/Select.java (at line 284)
	".//option[normalize-space(.) = " + Quotes.escape(text) + "]"));
	                                    ^^^^^^
Quotes cannot be resolved
----------
5 problems (5 errors)go aborted!
org.apache.tools.ant.BuildException: Compile failed; see the compiler error output for details.
```

**Comment by asolntsev** [comment]
I hope I fixed the compilation problem.

**Comment by asolntsev** [comment]
@andreastt I agree, we can regulate which tests to execute in build script. 

But compilation... Finding and compaling sources is minor compared to running tests. Build system could automatically find and compile all sources. It's just fast now days. It's not ok that I need to manually enumerate all sources in the build script.

Comon, guys, Selenium is the leader project for Automating The World, but it's own build system requires so much manual work!

**Comment by barancev** [comment]
*It's just fast now days.*

It is not about speed, build.desc files describe dependencies between parts of the code. This allows us to build various distribution artifacts from the single code tree and manage what class/resource files should be included into each of these artifacts, in five different programming languages.

Stop complaining about build system, it has its features for purpose :)

**Comment by andreastt** [comment]
I'm sorry for starting this digression, but I feel I need to address a few points:

> Build system could automatically find and compile all sources. It's just fast now days. It's not ok that I need to manually enumerate all sources in the build script.

We can't rely on IDE's to determine how sources are compiled.  The build system is primarily there, as @barancev points out, to provide us with reproducible and atomic builds.

Because Selenium is not one, single big project, we need the ability to build individual packages/jars.  For example the core APIs live in a jar that's separate to their implementation.  This means we can use the interfaces, without pulling in their various implementations, when browser vendors do their own implementations that typically live outside of our repository.  One example of this is the [Presto OperaDriver](https://github.com/operasoftware/operaprestodriver).

> Comon, guys, Selenium is the leader project for Automating The World, but it's own build system requires so much manual work!

Since mashing everything up in one single namespace isn't the right solution, defining the dependencies and relationships between different parts of the source tree is inevitable.

**Comment by andreastt** [comment]
In other news, your update passes the build! 👍

**Comment by andreastt** [comment]
Integrated in c2c6b4302a389ba74d4323d976be2ada6ed16c6c.

I forgot to ask you if you'd signed the CLA?

**Comment by asolntsev** [comment]
Yes, I signed CLA, and I already 
did pull requests before.

Sent from my HTC

----- Reply message -----

**Comment by andreastt** [comment]
Thanks for the patch, @asolntsev!

**Comment by asolntsev** [comment]
Yuhuu! 
@barancev @andreastt andreastt Sorry, I don't want to complain. My only goal is to suggest to make it simplier.
I understand that the build system is created to make multiple artifacts from the single code base. But it has few disadvantages. I believe we could improve them.

1. I need to manually list all the source files in build script.
2. There could be unused/uncompilable files in project. When I opened Selenium project in IDE two days ago, it was red. It's because IDE compiles all sources. 

I suggest the following. Build system could compile all the sources, and only at packaging phase define which jars should contain which packages. It could be done with wildcards, so I don't need to list all the source files. Something like that:

```
selenium-core.jar 
   + org/openqa/selenium/logging/**/*
   + org/openqa/selenium/support/**/*

selenium-chromedriver.jar:
   + org/openqa/selenium/chrome/**/*

selenium-chromedriver.jar:
   + org/openqa/selenium/firefox/**/*
```

**Pull request closed without merging**